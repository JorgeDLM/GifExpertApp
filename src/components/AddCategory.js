import React, { useState } from 'react'

export const AddCategory = ({categorias, setCategorias}) =>{

    // Hooks:
    const [inputValue, setInputValue] = useState('')
    const [mostrarAlerta, setMostrarAlerta] = useState({input:true})
    const [categoriaRepetida, setCategoriaRepetida] = useState({categoria: true})

    // Funciones
    const convertirMayusculas = (e) => e.charAt(0) === ' ' ? e.charAt(1).toUpperCase() + e.toLowerCase().slice(1) : e.charAt(0).toUpperCase() + e.toLowerCase().slice(1)

    const handleAgregar = () => {
        setCategorias([convertirMayusculas(inputValue), ...categorias])
    }

    const sinCambio = inputValue !== '' && inputValue !== ' '
    
    // para saber si hay repetidas
    const contieneTextoRepetido = categorias.indexOf(inputValue) === -1
    // para sabir si hay repetido y aun no hay cambios ene l iput
    const sinCambioTextoRepetido = sinCambio && contieneTextoRepetido

    // input en ENTER se activa:
    const handleSubmit = (e) => {
        // esto lo que hace es evitar el funcionamiento por defecto del form:
        e.preventDefault();

        // utilizando el form al picar enter en el input se activa el submit y me evito este codigo:
        // e => (e.keyCode === 13)
    }

    
    const handleChange = (e) => {
        // manejandoDobleEspacio
        setInputValue(e.target.value.replace(/ +(?= )/g,''))
    }
    
    const agregarCategoria = () => {
        (sinCambioTextoRepetido) && handleAgregar(); 
        (sinCambioTextoRepetido) && setMostrarAlerta(false); 
        (sinCambio && !contieneTextoRepetido) && setCategoriaRepetida(false); 
        (sinCambio && contieneTextoRepetido) && setCategoriaRepetida(true)
    }
    
    // styles
    const boton = {backgroundColor: 'white', borderRadius: '10px', paddingBottom: '2px', marginLeft: '5px'}
    const alerta = {fontSize: '12px', paddingTop: '3px',}
    
    return(
        <>
            <form onSubmit={handleSubmit}>
                <input 
                    maxLength={20} 
                    type="text" 
                    placeholder="Buscar GIF's" 
                    value={convertirMayusculas(inputValue)} 
                    // ya no es necesario porque esta el form onEnter:
                    // onKeyDown={e => (e.keyCode === 13) && agregarCategoria()} 
                    onChange={handleChange} 
                />
                <button onClick={agregarCategoria}  style={{...boton}}>Agregar</button>
            </form>

           {(!mostrarAlerta && categoriaRepetida) && 
                <div style={{...alerta, color: 'green'}}>
                    Categoria {categorias[categorias.length -1 ]} creada con exito!
                </div>
           }

            {!(categoriaRepetida) && 
                <div style={{...alerta, color: 'red'}}>Categoria repetida.</div>
            }
        </>
    )
}



export default AddCategory;