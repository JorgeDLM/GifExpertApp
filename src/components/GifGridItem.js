import React from 'react'
// para importar propTypes rapido: imp tab
import PropTypes from 'prop-types'

export const GifGridItem = ({title, url, id, ultimaCategoria, categorias}) => {
    return (
        <div className={"card"}>
            <img style={{}} src={url} alt={title}/>
            <p style={{fontWeight: 'bold'}}>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}

GifGridItem.defaultValue = {
    title: 'Hola'
}


export default GifGridItem