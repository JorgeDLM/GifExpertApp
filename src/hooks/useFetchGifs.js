import { useState, useEffect } from "react"
import { getGifs } from "../helpers/getGifs"

export const useFetchGifs = ( categorias, cantidad ) => {
    // el apikey lo quiero pedir aqui y no en GifGrid: entonces en props no lo mando pero si lo regreso en getGifs
    const apiKey = '6ndyzTJHSlDiGxgJKJkO1spzFSyEFRD6'
    const [state, setState] = useState({
        data: [], 
        loading: true
    })
    // el useEffect: Es el equivalente funcional a componentDidMount, componentDidUpdate, y componentWillUnmount en los componentes de clase.
    useEffect(() => {
        getGifs( categorias, cantidad, apiKey )
            .then( imgs => {
                setState({
                    data: imgs,
                    loading: false
                })
            })
        },[categorias, cantidad, apiKey])
        
    // asi podria activar despues de un timpo un cambio de estado
    // setTimeout(() => {
    //     setState({loading: false})
    // }, 3000);

    return state
}
