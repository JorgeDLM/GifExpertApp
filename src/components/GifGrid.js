import React from 'react'
import GifGridItem from './GifGridItem'
// import { getGifs } from '../helpers/getGifs'
import { useFetchGifs } from '../hooks/useFetchGifs'

export const GifGrid = ({categorias, cantidad, ultimaCategoria, pruebas}) => {


    // Cuando esto se ejecuta lo que hace es que hace que se vuelva a ejecutar el getGifs cada que haya un cambio, esto es porque hay un cambio en este nivel en el componente
    // const [count, setCount] = useState(0);

    // const [images, setImages] = useState([])

    // commente esto porque voy a hacer un custom hook, los hooks siempre empiezan con useFuncionAlgo esta en carpeta de hooks
    // useEffect(() => {
    //     getGifs(categorias, cantidad, apiKey).then(img => setImages(img));

    //     // el segundo argumento es un arreglo de dependencias, sirve para parar el que se ejecute mas el getGifs, esta parte es igual al componentDidMount()
    // }, [categorias, cantidad])

    // const state = useFetchGifs()
    // lo imsmo pero desestructurado
    // para darle un nombre especifico a la data se usan : data:images
    const { data:images,loading } = useFetchGifs(categorias, cantidad)
    

    // si esto lo pongo aqui se ejecuta cuando carga
    // getGifs()
    // lo pase arriba a useEffect para que no se ejecute el codigo cada vez que haya un cambio en el componente
    return (
            <>
            {pruebas && <h2>{images.length}</h2>}
                <div style={{textAlign: 'center', marginBottom: '4px', fontSize: '25px', fontWeight: 'bolder'}}>{categorias}</div>
                {loading && <p className="animate__flash">
                    Cargando...
                </p> }
                    <div className="card-grid">
                        {/* <button onClick={() => setCount(count +1)}>Set count +1</button> */}
                        {images.reverse().map(image =>
                            <GifGridItem
                            ultimaCategoria={ultimaCategoria}
                            categorias={categorias}
                            lastItem={[categorias[1]] && true}
                            key={image.id}
                            // esto es como yo lo hacia
                            // images={image}
                            // id={id} etc...
                            // esto es como debe de ser:
                            // mando un spread de la informacion de image como props, cada prop de las imagenes se manda como prop independiente
                            {...image}
                            />
                        )}
                    </div>
            </>
    )
}
