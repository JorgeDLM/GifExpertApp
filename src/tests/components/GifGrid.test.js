import React from 'react'

import { GifGrid } from "../../components/GifGrid"
import { shallow } from "enzyme"
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')

describe('pruebas en <GifGrid/>', () => {
    
    // const cantidad = 10
    const gifs = [{
        id: 'ABC',
        url: 'https://localhost/cualquier-imagen.jpg',
        title: 'abc'
    },
    {
        id: 'ABC2',
        url: 'https://localhost/cualquier-imagen.jpg',
        title: 'abc'
    }]
    
    // let wrapper = shallow(<GifGrid categorias="juan" cantidad={cantidad} />)
    // beforeEach(() => {
        //     wrapper = shallow(<GifGrid categorias="juan" cantidad={cantidad} />)
        // })
        
    test('Debe de renderizarse el Snapshot correctamente', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })
        const wrapper = shallow( <GifGrid categorias="juan"  /> )
       
        expect( wrapper ).toMatchSnapshot()
        
    })
    
    test('Debe de mostrar items cuando se cargan imagenes con useFetchGifs', () => {
        // voy a crear un mock para fingir que ya llego la data -> esta rriba jest.mock('ruta a hook')
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        })
        const wrapper = shallow(<GifGrid categorias="juan"  />)
       
        // expect( wrapper ).toMatchSnapshot()

        // el p en este caso contiene el cargando (como no esta cargando entonces false):
        expect( wrapper.find('p').exists() ).toBe(false)

        // como se crea un GifGridItem por gifs, entonces:
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length )


    })
    
})
