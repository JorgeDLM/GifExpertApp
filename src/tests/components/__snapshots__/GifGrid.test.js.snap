// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <GifGrid/> Debe de renderizarse el Snapshot correctamente 1`] = `
<Fragment>
  <div
    style={
      Object {
        "fontSize": "25px",
        "fontWeight": "bolder",
        "marginBottom": "4px",
        "textAlign": "center",
      }
    }
  >
    juan
  </div>
  <p
    className="animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
